File: output/CodeLlama/avatar/Python/Java/atcoder_AGC017_A.java
import java.util.Scanner;
public class atcoder_AGC017_A {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int P = sc.nextInt();
        int ans = 0;
        int[] lis = new int[N];
        for (int i = 0; i < N; i++) {
            lis[i] = sc.nextInt();
        }
        int[] ls = new int[N];
        for (int a : lis) {
            ls[a % 2]++;
        }
        int one = ls[1];
        int zero = ls[0];
        int pattern_a = 0;
        int pattern_b = 0;
        for (int j = 0; j <= zero; j++) {
            pattern_b += combi(zero, j);
        }
        int time = 0;
        while (time <= one) {
            if (time % 2 == P) {
                pattern_a += combi(one, time);
            }
            time++;
        }
        System.out.println((int) (pattern_a * pattern_b));
    }
    public static int combi(int N, int K) {
        int a = 1;
        for (int i = 0; i < K; i++) {
            a *= N - i;
        }
        for (int j = 0; j < K; j++) {
            a /= j + 1;
        }
        return a;
    }
}

File: output/CodeGen/avatar/Java/Python/atcoder_AGC017_A.py
def C(n,m):
    res = 1
    for i in range(m-n+1,m+1):
        res = res * i
    for i in range(1,n+1):
        res = res / i
    return res

def main():
    sc = Scanner()
    n, p = sc.nextInt(), sc.nextInt()
    k0, k1 = 0, 0
    for i in range(n):
        x = sc.nextInt()
        if x % 2 == 1:
            k1 += 1
        else:
            k0 += 1
    ans = 1
    for i in range(k0):
        ans = ans * 2
    if p == 0:
        for i in range(k1):
            ans = ans * 2
            for j in range(i, k1):
                ans = ans + C(i, j)
    else:
        for i in range(k1):
            ans = ans * 2
            for j in range(i+1, k1):
                ans = ans + C(i, j)
    print(ans)

main()